src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,387 @@
+        if (!backEdgesEnabled) {
+            BackEdges.enable(g);
+        }
+
+        boolean hadChange = false;
+        boolean changes;
+        int maxChanges = 0;
+
+        do {
+            changes = false;
+            nodeValues.clear();
+            dominateMap.clear();
+            worklist = new DefaultWorklist<>();
+
+            CSEVisitor visitor = new CSEVisitor();
+
+
+            g.walkTopological(new WorklistFiller(worklist));
+            while (!worklist.isEmpty()) {
+                Node n = worklist.dequeue();
+                n.accept(visitor);
+            }
+
+            for (var replacement : nodeValues.entrySet()) {
+                // for each replacement merge nodes together
+                Node orig = replacement.getKey();
+                Node n = replacement.getValue();
+
+                changes |= transform(g, orig, n, orig.getOpCode());
+            }
+            if (changes) {
+                hadChange = true;
+            }
+            maxChanges++;
+        } while (changes && maxChanges < 10);
+
+        if (!backEdgesEnabled) {
+            BackEdges.disable(g);
+        }
+        binding_irgopt.remove_bads(g.ptr);
+        binding_irgopt.remove_unreachable_code(g.ptr);
+        binding_irgopt.remove_bads(g.ptr);
+
+        return hadChange;
+    }
+
+    /**
+     * Transform the given node with the given associated lattice element if
+     * that element is constant.
+     */
+    private boolean transform(Graph g, Node orig, Node replacement, binding_irnode.ir_opcode opcode) {
+        Node newNode = switch (opcode) {
+            case iro_Add -> {
+                Add add = (Add) orig;
+                yield g.newAdd(orig.getBlock(), add.getLeft(), add.getRight());
+            }
+            case iro_And -> {
+                And and = (And) orig;
+                yield g.newAnd(orig.getBlock(), and.getLeft(), and.getRight());
+            }
+            case iro_Or -> {
+                Or or = (Or) orig;
+                yield g.newOr(orig.getBlock(), or.getLeft(), or.getRight());
+            }
+            case iro_Sub -> {
+                Sub sub = (Sub) orig;
+                yield g.newSub(orig.getBlock(), sub.getLeft(), sub.getRight());
+            }
+            case iro_Eor -> {
+                Eor eor = (Eor) orig;
+                yield g.newEor(orig.getBlock(), eor.getLeft(), eor.getRight());
+            }
+            case iro_Mul -> {
+                Mul mul = (Mul) orig;
+                yield g.newMul(orig.getBlock(), mul.getLeft(), mul.getRight());
+            }
+            case iro_Minus -> {
+                Minus minus = (Minus) orig;
+                yield g.newMinus(orig.getBlock(), minus.getOp());
+            }
+            case iro_Const -> {
+                Const cons = (Const) orig;
+                yield g.newConst(cons.getTarval().asInt(), cons.getMode());
+            }
+            case iro_Sel -> {
```

Comment:
I think out optimizations runs after `lower`, so we shouldn't have `Sel` nodes at this point.